test/functional/mempool_updatefromblock.py

```diff
@@ -52,42 +49,17 @@ def transaction_graph_test(self, size, n_tx_to_mine=None, start_input_txid='', e
             self.log.debug('Preparing transaction #{}...'.format(i))
             # Prepare inputs.
             if i == 0:
-                inputs = [{'txid': start_input_txid, 'vout': 0}]
-                inputs_value = self.nodes[0].gettxout(start_input_txid, 0)['value']
+                parent_utxo = miniwallet.get_utxo()
             else:
-                inputs = []
-                inputs_value = 0
-                for j, tx in enumerate(tx_id[0:i]):
-                    # Transaction tx[K] is a child of each of previous transactions tx[0]..tx[K-1] at their output K-1.
-                    vout = i - j - 1
-                    inputs.append({'txid': tx_id[j], 'vout': vout})
-                    inputs_value += self.nodes[0].gettxout(tx, vout)['value']
-
-            self.log.debug('inputs={}'.format(inputs))
-            self.log.debug('inputs_value={}'.format(inputs_value))
-
-            # Prepare outputs.
-            tx_count = i + 1
-            if tx_count < size:
-                # Transaction tx[K] is an ancestor of each of subsequent transactions tx[K+1]..tx[N-1].
-                n_outputs = size - tx_count
-                output_value = ((inputs_value - fee) / Decimal(n_outputs)).quantize(Decimal('0.00000001'))
-                outputs = {}
-                for _ in range(n_outputs):
-                    outputs[self.nodes[0].getnewaddress()] = output_value
-            else:
-                output_value = (inputs_value - fee).quantize(Decimal('0.00000001'))
-                outputs = {end_address: output_value}
+                parent_utxo = miniwallet.get_utxo(txid=tx_id[i-1]['txid'])
 
-            self.log.debug('output_value={}'.format(output_value))
-            self.log.debug('outputs={}'.format(outputs))
+            tx = miniwallet.send_self_transfer(from_node=self.nodes[0], utxo_to_spend=parent_utxo, sequence=SEQUENCE_FINAL)
```

Comment:
Why does this need `SEQUENCE_FINAL`?

Reply:
Without `SEQUENCE_FINAL`, the test broke at `assert_equal(len(node.getrawmempool()), size)` after the first block is invalidated, as the mempool has 26 transactions instead of the expected 100.

In a simpler example with fewer transactions without `SEQUENCE_FINAL`, I can see after the block is invalidated, transactions are removed from the mempool because of this place: https://github.com/bitcoin/bitcoin/blob/f05cf59d91eb03857dd9bdcc77607764da0349d2/src/consensus/tx_verify.cpp#L53